packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,415 @@
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instanceDefinition: Definition
+) {
+    setPrototypeOf(instance, instanceDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instanceDefinition);
+    // attributes patches when needed
+    if (instanceDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instanceDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    new instanceDefinition.UserCtor();
+
+    patchAttributesDuringUpgrade(instance, registeredDefinition, instanceDefinition);
+}
+
+function getDefinitionForConstructor(constructor: CustomElementConstructor): Definition {
+    if (!isFunction(constructor) || !isObject(constructor.prototype)) {
+        throw new TypeError('The referenced constructor is not a constructor.');
+    }
+    let definition = definitionForConstructor.get(constructor);
+    if (isUndefined(definition)) {
+        definition = createDefinitionRecord(constructor);
+        definitionForConstructor.set(constructor, definition);
+    }
+    return definition;
+}
+
+export function patchCustomElementRegistry() {
+    const { customElements: nativeRegistry } = window;
+    const { define: nativeDefine, whenDefined: nativeWhenDefined, get: nativeGet } = nativeRegistry;
+
+    // patch for the global registry define mechanism
+    CustomElementRegistry.prototype.define = function define(
+        this: CustomElementRegistry,
+        tagName: string,
+        constructor: CustomElementConstructor,
+        options?: ElementDefinitionOptions
+    ): void {
+        if (options && options.extends) {
+            // TODO [#2877]: should we support `extends`?
+            throw new DOMException(
+                'NotSupportedError: "extends" key in customElements.define() options is not supported.'
+            );
+        }
+        nativeGet.call(this, tagName); // SyntaxError if The provided name is not a valid custom element name.
+        if (!isUndefined(globalDefinitionsByTag.get(tagName))) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': the name "${tagName}" has already been used with this registry`
+            );
+        }
+        const definition = getDefinitionForConstructor(constructor);
+        if (!isUndefined(globalDefinitionsByClass.get(constructor))) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': this constructor has already been used with this registry`
+            );
+        }
+        globalDefinitionsByTag.set(tagName, definition);
```

Comment:
This bug was introduced by @nolanlawson today by removing the following line:

> nativeGet.call(this, tagName); // SyntaxError if The provided name is not a valid custom element name.

Which was prior to this line to guarantee that the name was a valid name, and the constructor is always a valid constructor because it is the pivot that we build. I'm fine going with this approach from @pmdartus though.

Reply:
Good catch. Fixed and added some tests (for both vanilla `customElements.define()` as well as `lwc.createElement()`).

Reply:
Just saw Caridy's comment. Yeah, let's avoid calling `customElements.get()` inside of `define()` anyway. We can let `nativeDefine` throw the error.

In any case, we also need to handle this in `createPivotingClass`, so we can use the same strategy for both.